### PR TITLE
Video Player Refactor

### PR DESCRIPTION
**Pull request by sorenhan**
Implemented a cleaner video player interface to replace the current AltspaceVR-exclusive RPC-based one.

This is now complete. 
This is a breaking change, in that old MREs would not play sounds (as payload has changed and been renamed)

### CODE REVIEW

File: MREUnityRuntime/MREUnityRuntimeLib/Assets/Definitions.cs [hunk 1]
```diff
@@ -98,8 +103,31 @@ public enum AlphaMode
     }
 
     /// <summary>
-    /// Contains material asset info
+    /// Type  describing how to interpret the video Source URI
     /// </summary>
+    public enum VideoSourceType
```
Comment (eanders-MS) [inline MREUnityRuntime/MREUnityRuntimeLib/Assets/Definitions.cs]: >VideoSourceType [](start = 16, length = 15)

Noting here that this should be removed, to match the SDK implementation.

File: MREUnityRuntime/MREUnityRuntimeLib/Core/SoundManager.cs [hunk 2]
```diff
@@ -114,6 +113,10 @@ public void ApplySoundStateOptions(Actor actor, AudioSource soundInstance, Sound
                     soundInstance.minDistance = options.RolloffStartDistance.Value;
                     soundInstance.maxDistance = options.RolloffStartDistance.Value * 1000000.0f;
                 }
+                if (options.Time != null)
+                {
+                    soundInstance.time=options.Time.Value;
```
Comment (eanders-MS) [inline MREUnityRuntime/MREUnityRuntimeLib/Core/SoundManager.cs]: >me=op [](start = 36, length = 5)

cslint: Whitespace around `=`

File: MREUnityRuntime/MREUnityRuntimeLib/Messaging/Payloads/PayloadTypeRegistry.cs [hunk 3]
```diff
@@ -43,7 +43,7 @@ namespace MixedRealityExtension.Messaging.Payloads
     [PayloadType(typeof(SetAnimationState), "set-animation-state")]
     [PayloadType(typeof(SetAuthoritative), "set-authoritative")]
     [PayloadType(typeof(SetBehavior), "set-behavior")]
-    [PayloadType(typeof(SetSoundState), "set-sound-state")]
```
Comment (eanders-MS) [inline MREUnityRuntime/MREUnityRuntimeLib/Messaging/Payloads/PayloadTypeRegistry.cs]: I think we're fine with breaking back compatibility here, but want to make sure we're in agreement that this is a breaking change.

File: MREUnityRuntime/MREUnityRuntimeLib/Core/SoundManager.cs (hunk 2 above)
Comment (sorenhan) [reply MREUnityRuntime/MREUnityRuntimeLib/Core/SoundManager.cs]: fixed

File: MREUnityRuntime/MREUnityRuntimeLib/Assets/Definitions.cs (hunk 1 above)
Comment (sorenhan) [reply MREUnityRuntime/MREUnityRuntimeLib/Assets/Definitions.cs]: has been removed

File: MREUnityRuntime/MREUnityRuntimeLib/Messaging/Payloads/PayloadTypeRegistry.cs (hunk 3 above)
Comment (sorenhan) [reply MREUnityRuntime/MREUnityRuntimeLib/Messaging/Payloads/PayloadTypeRegistry.cs]: I will call this out in SDK channel.

File: MREUnityRuntime/MREUnityRuntimeLib/Assets/AssetLoader.cs [hunk 4]
```diff
@@ -363,7 +386,10 @@ internal async void OnCreateAsset(CreateAsset payload, Action onCompleteCallback
             }
             else
             {
-                response.FailureMessage = $"Not implemented: CreateAsset of type {(def.Prefab != null ? "Prefab" : "Mesh")}";
+                if (response.FailureMessage == null)
+                {
+                    response.FailureMessage = $"Not implemented: CreateAsset of new asset type";
```
Comment (stevenvergenz) [inline MREUnityRuntime/MREUnityRuntimeLib/Assets/AssetLoader.cs]: This error message seems a little vague, but I can't come up with anything better.

File: MREUnityRuntime/MREUnityRuntimeLib/Core/Actor.cs [hunk 5]
```diff
@@ -1235,32 +1235,53 @@ private void OnSetAnimationState(SetAnimationState payload, Action onCompleteCal
...
                 AudioSource soundInstance = App.SoundManager.TryAddSoundInstance(this, payload.Id, payload.SoundAssetId, payload.Options, payload.StartTimeOffset);
                 if (soundInstance)
                 {
-                    if (_soundInstances == null)
+                    _mediaInstances.Add(payload.Id, soundInstance);
+                }
+                else
+                {
+                    var factory = MREAPI.AppsAPI.VideoPlayerFactory
+                        ?? throw new ArgumentException("Cannot movie player not implemented library.");
```
Comment (stevenvergenz) [inline MREUnityRuntime/MREUnityRuntimeLib/Core/Actor.cs]: The semiotics of this error message are beyond me.

**Comment by stevenvergenz** [comment]
>         public Guid SoundAssetId { get; set; }

Rename this field to be generic. `MediaAssetId`?

---
Refers to: MREUnityRuntime/MREUnityRuntimeLib/Messaging/Payloads/NetworkCommandPayloads.cs:329 in 5e3bf32. [](commit_id = 5e3bf32b674fb0745dcc3cc2fd2a43b2f0419012, deletion_comment = False)

File: MREUnityRuntime/MREUnityRuntimeLib/Core/Actor.cs [hunk 6]
```diff
@@ -1235,32 +1235,53 @@ private void OnSetAnimationState(SetAnimationState payload, Action onCompleteCal
...
+                        videoPlayer.Play(videoStreamDescription, payload.Options, payload.StartTimeOffset);
                     }
-                    _soundInstances.Add(payload.Id, soundInstance);
+                    _mediaInstances.Add(payload.Id, videoPlayer);
                 }
             }
             else
             {
-                if (_soundInstances != null && _soundInstances.TryGetValue(payload.Id, out AudioSource soundInstance))
+                if (_mediaInstances != null && _mediaInstances.TryGetValue(payload.Id, out System.Object mediaInstance))
```
Comment (stevenvergenz) [inline MREUnityRuntime/MREUnityRuntimeLib/Core/Actor.cs]: cslint: Combine into `else if` to reduce indentation.

File: MREUnityRuntime/MREUnityRuntimeLib/Core/Actor.cs [hunk 7]
```diff
@@ -1235,32 +1235,53 @@ private void OnSetAnimationState(SetAnimationState payload, Action onCompleteCal
...
                 {
-                    switch (payload.SoundCommand)
+                    switch (payload.MediaCommand)
```
Comment (stevenvergenz) [inline MREUnityRuntime/MREUnityRuntimeLib/Core/Actor.cs]: There are only two states, why a switch statement?

File: MREUnityRuntime/MREUnityRuntimeLib/Core/Actor.cs [hunk 8]
```diff
@@ -1270,21 +1291,31 @@ private void OnSetSoundState(SetSoundState payload, Action onCompleteCallback)
 
         public bool CheckIfSoundExpired(Guid id)
```
Comment (stevenvergenz) [inline MREUnityRuntime/MREUnityRuntimeLib/Core/Actor.cs]: >CheckIfSoundExpired [](start = 20, length = 19)

Combine into conditional access: `_mediaInstances?.TryGetValue`

File: MREUnityRuntime/MREUnityRuntimeLib/Core/Actor.cs (hunk 6 above)
Comment (stevenvergenz) [inline MREUnityRuntime/MREUnityRuntimeLib/Core/Actor.cs]: Combine into conditional access: `_mediaInstances?.TryGetValue`

File: MREUnityRuntime/MREUnityRuntimeLib/Core/Actor.cs [hunk 9]
```diff
@@ -1270,21 +1291,31 @@ private void OnSetSoundState(SetSoundState payload, Action onCompleteCallback)
...
         {
-            if (_soundInstances != null && _soundInstances.TryGetValue(id, out AudioSource soundInstance))
+            if (_mediaInstances != null && _mediaInstances.TryGetValue(id, out System.Object mediaInstance))
```
Comment (stevenvergenz) [inline MREUnityRuntime/MREUnityRuntimeLib/Core/Actor.cs]: Combine into conditional access: `_mediaInstances?.TryGetValue`

**Comment by stevenvergenz** [comment]
>         public float? Pitch;

Does this apply to videos?

---
Refers to: MREUnityRuntime/MREUnityRuntimeLib/Core/MediaStateOptions.cs:13 in 5e3bf32. [](commit_id = 5e3bf32b674fb0745dcc3cc2fd2a43b2f0419012, deletion_comment = False)

File: MREUnityRuntime/MREUnityRuntimeLib/Core/SoundManager.cs [hunk 10]
```diff
@@ -35,14 +35,13 @@ public SoundManager(MixedRealityExtensionApp app)
 
         #region Public Methods
 
-        public AudioSource TryAddSoundInstance(Actor actor, Guid id, Guid soundAssetId, SoundStateOptions options, float? startTimeOffset)
+        public AudioSource TryAddSoundInstance(Actor actor, Guid id, Guid soundAssetId, MediaStateOptions options, float? startTimeOffset)
         {
-            var obj = MREAPI.AppsAPI.AssetCache.GetAsset(soundAssetId);
             var audioClip = MREAPI.AppsAPI.AssetCache.GetAsset(soundAssetId) as AudioClip;
             if (audioClip != null)
```
Comment (stevenvergenz) [inline MREUnityRuntime/MREUnityRuntimeLib/Core/SoundManager.cs]: nit: If preferred, return early to reduce indentation

**Comment by sorenhan** [comment]
>         public float? Pitch;

Maybe it's not that logical terminology, but it technically works. we may want to iterate on this later

---
In reply to: [510561044](https://github.com/Microsoft/mixed-reality-extension-unity/pull/86#issuecomment-510561044) [](ancestors = 510561044)

---
Refers to: MREUnityRuntime/MREUnityRuntimeLib/Core/SoundStateOptions.cs:13 in 5e3bf32. [](commit_id = 5e3bf32b674fb0745dcc3cc2fd2a43b2f0419012, deletion_comment = False)

File: MREUnityRuntime/MREUnityRuntimeLib/Core/Actor.cs (hunk 9 above)
Comment (sorenhan) [reply MREUnityRuntime/MREUnityRuntimeLib/Core/Actor.cs]: I think it's less readable that way, so I'm leaving it the way it is.

---
In reply to: [302638875](https://github.com/Microsoft/mixed-reality-extension-unity/pull/86#discussion_r302638875) [](ancestors = 302638875)

File: MREUnityRuntime/MREUnityRuntimeLib/Core/Actor.cs (hunk 5 above)
Comment (sorenhan) [reply MREUnityRuntime/MREUnityRuntimeLib/Core/Actor.cs]: yeah, that was weird. 

---
In reply to: [302633494](https://github.com/Microsoft/mixed-reality-extension-unity/pull/86#discussion_r302633494) [](ancestors = 302633494)

File: MREUnityRuntime/MREUnityRuntimeLib/Core/Actor.cs (hunk 7 above)
Comment (sorenhan) [reply MREUnityRuntime/MREUnityRuntimeLib/Core/Actor.cs]: I refactored this into a cleaner 3-item switch statement, so fixed.

---
In reply to: [302636501](https://github.com/Microsoft/mixed-reality-extension-unity/pull/86#discussion_r302636501) [](ancestors = 302636501)

File: MREUnityRuntime/MREUnityRuntimeLib/Core/Actor.cs (hunk 6 above)
Comment (sorenhan) [reply MREUnityRuntime/MREUnityRuntimeLib/Core/Actor.cs]: I refactored this into a cleaner 3-level switch statement, so fixed.

---
In reply to: [302635996](https://github.com/Microsoft/mixed-reality-extension-unity/pull/86#discussion_r302635996) [](ancestors = 302635996)

**Pull request merged**